lib/luaxtable.c

```diff
@@ -32,48 +14,170 @@
 	};
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+static struct {
+	lunatik_object_t *target;
+	lunatik_object_t *match;
+	bool match_fallback;
+	unsigned int target_fallback;
+} luaxtable_hooks = {NULL, NULL, false, XT_CONTINUE};
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, const char *op, int nargs, int nret, int fallback)
 {
-	return true;
+	int base = lua_gettop(L) - nargs;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("%s: could not find ops table\n", op);
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	lua_insert(L, base + 1); /* op */
+	lua_pop(L, 1); /* table */
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		goto err;
+	}
+	if (nret == 1) {
+		int ret = luaL_optinteger(L, -1, 0);
+		ret = ret < 0 ? fallback : ret;
+		return ret;
+	}
+err:
+	return fallback;
 }
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static inline lunatik_object_t *luaxtable_getskb(lua_State *L, luaxtable_t *xtable)
 {
-	return 0;
+	if (lunatik_getregistry(L, xtable->skb) != LUA_TUSERDATA)
+		return NULL;
+
+	return (lunatik_object_t *)lunatik_toobject(L, -1);
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+static inline void luaxtable_pushparams(lua_State *L, const struct xt_action_param *par)
 {
+	lua_newtable(L);
+	lua_pushboolean(L, par->hotdrop);
+	lua_setfield(L, -2, "hotdrop");
+	lua_pushinteger(L, par->thoff);
+	lua_setfield(L, -2, "thoff");
+	lua_pushinteger(L, par->fragoff);
+	lua_setfield(L, -2, "fragoff");
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 10, 0))
+	lua_pushinteger(L, xt_hooknum(par));
+#else
+	lua_pushinteger(L, par->hooknum);
+#endif
+	lua_setfield(L, -2, "hooknum");
+	lua_pushvalue(L, -1); /* param table */
+	lua_insert(L, lua_gettop(L) - 2); /* stack: param, skb, param */
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
+static inline int luaxtable_pushskb(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, uint8_t opt)
 {
-	return NF_ACCEPT;
+	lunatik_object_t *data = luaxtable_getskb(L, xtable);
+	if (data == NULL) {
+		pr_err("could not find skb\n");
+		return -1;
+	}
+
+	luadata_reset(data, skb->data, skb_headlen(skb), opt);
+	return 0;
 }
 
-static int luaxtable_target_check(const struct xt_tgchk_param *par)
+static int luaxtable_domatch(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, struct xt_action_param *par)
 {
-	return 0;
+	int ret;
+	if (luaxtable_pushskb(L, xtable, skb, LUADATA_OPT_READONLY) == -1)
+		return luaxtable_hooks.match_fallback;
+	luaxtable_pushparams(L, par);
+
+	ret = luaxtable_invoke(L, xtable, "match", 2, 1, luaxtable_hooks.match_fallback);
+	lua_getfield(L, -2, "hotdrop");
+	par->hotdrop = lua_toboolean(L, -1);
```

Comment:
btw, why can't we set `hotdrop` on target?

Reply:
Because target callback sends it's arguments as `const xt_action_param`, so we are not allowed to set `hotdrop` here, though we can force it, but that is not intended. `hotdrop` is generally used when the packet qualifies for that particular module, but is not in the correct format or somehow corrupt. For example, `conntrack` uses hotdrop to drop packets if it was not able to get memory using `kmalloc`.